src/System.Buffers/src/System/Buffers/Utilities.cs

```diff
@@ -8,19 +8,37 @@ namespace System.Buffers
 {
     internal static class Utilities
     {
+        private static readonly int[] s_logTable256 = new int[256] {
+           -1, 0, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3,
+            4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
+            5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5,
+            5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5,
+            6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6,
+            6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6,
+            6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6,
+            6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6,
+            7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
+            7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
+            7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
+            7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
+            7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
+            7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
+            7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
+            7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
+        };
+
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         internal static int SelectBucketIndex(int bufferSize)
         {
-            uint bitsRemaining = ((uint)bufferSize - 1) >> 4;
-            int poolIndex = 0;
-
-            while (bitsRemaining > 0)
-            {
-                bitsRemaining >>= 1;
-                poolIndex++;
-            }
+            uint input = ((uint)bufferSize - 1) >> 4;
```

Comment:
The statics and array access do not have exactly same characteristics in C# and C++ (you pay for range check and cctor trigger). I would expect the following implementation to be better - in particular when the code is precompiled and the ctor trigger is a bit more expensive.

```
static int SelectBucketIndex(int bufferSize)
{
    uint bitsRemaining = ((uint)bufferSize - 1) >> 4;

    int poolIndex = 0;
    if (bitsRemaining > 0xFFFF) { bitsRemaining >>= 16; poolIndex = 16; }
    if (bitsRemaining > 0xFF) { bitsRemaining >>= 8; poolIndex += 8; }
    if (bitsRemaining > 0xF) { bitsRemaining >>= 4; poolIndex += 4; }
    if (bitsRemaining > 0x3) { bitsRemaining >>= 2; poolIndex += 2; }
    if (bitsRemaining > 0x1) { bitsRemaining >>= 1; poolIndex += 1; }
    return poolIndex + (int)bitsRemaining;
}
```